### PR TITLE
Use run loop for updating component helper's component 

### PR DESCRIPTION
**Pull request by lukemelia**
This commit addresses a failing test prepared by @bagby. The test illustrates a race between streams where one stream triggers the component helper to create a new component instance, and the other stream updates a binding provided to that component helper.

The solution in this PR is to schedule the component helper's "swap" on the run loop, so that by the time it executes, other bound streams to catch up before rendering a new component.

Would love input from @mmun, @krisselden and/or @mixonic, and also confirmation from @bagby that this addresses your real-world scenario.

### CODE REVIEW

File: packages/ember-htmlbars/tests/helpers/component_test.js [hunk 1]
```diff
@@ -25,6 +27,49 @@ if (Ember.FEATURES.isEnabled('ember-htmlbars-component-helper')) {
     }
   });
 
+  test("component helper with bound properties are updating correctly in init of component", function() {
+    registry.register('component:foo-bar', Ember.Component.extend({
```
Comment (rwjblue) [inline packages/ember-htmlbars/tests/helpers/component_test.js]: Should import `Component`.

File: packages/ember-htmlbars/tests/helpers/component_test.js [hunk 2]
```diff
@@ -3,9 +3,11 @@ import Registry from "container/registry";
 import EmberView from "ember-views/views/view";
 import compile from "ember-template-compiler/system/compile";
 import { runAppend, runDestroy } from "ember-runtime/tests/utils";
+import { computed } from "ember-metal/computed";
 
 var set = Ember.set;
```
Comment (rwjblue) [inline packages/ember-htmlbars/tests/helpers/component_test.js]: Should import this.

File: packages/ember-htmlbars/tests/helpers/component_test.js [hunk 3]
```diff
@@ -3,9 +3,11 @@ import Registry from "container/registry";
...
 var get = Ember.get;
+var on  = Ember.on;
```
Comment (rwjblue) [inline packages/ember-htmlbars/tests/helpers/component_test.js]: Should import this.

File: packages/ember-htmlbars/tests/helpers/component_test.js [hunk 4]
```diff
@@ -3,9 +3,11 @@ import Registry from "container/registry";
 import EmberView from "ember-views/views/view";
 import compile from "ember-template-compiler/system/compile";
 import { runAppend, runDestroy } from "ember-runtime/tests/utils";
+import { computed } from "ember-metal/computed";
 
 var set = Ember.set;
 var get = Ember.get;
```
Comment (rwjblue) [inline packages/ember-htmlbars/tests/helpers/component_test.js]: Should import this.

File: packages/ember-htmlbars/tests/helpers/component_test.js [hunk 5]
```diff
@@ -25,6 +27,49 @@ if (Ember.FEATURES.isEnabled('ember-htmlbars-component-helper')) {
     }
   });
 
+  test("component helper with bound properties are updating correctly in init of component", function() {
+    registry.register('component:foo-bar', Ember.Component.extend({
+      assertOnInit: on('init', function() {
+        equal(get(this, 'location'), 'Caracas', 'location is bound on init');
+      })
+    }));
+    registry.register('component:baz-qux', Ember.Component.extend({
```
Comment (rwjblue) [inline packages/ember-htmlbars/tests/helpers/component_test.js]: Should import `Component`.

File: packages/ember-htmlbars/tests/helpers/component_test.js [hunk 6]
```diff
@@ -25,6 +27,49 @@ if (Ember.FEATURES.isEnabled('ember-htmlbars-component-helper')) {
...
+        }
+      }),
+      location: 'Caracas',
+      template: compile('{{#component view.dynamicComponent location=view.location}}arepas!{{/component}}')
+    }).create();
+
+    runAppend(view);
+    equal(view.$().text(), 'yippie! Caracas arepas!', 'component was looked up and rendered');
+
+    Ember.run(function() {
```
Comment (rwjblue) [inline packages/ember-htmlbars/tests/helpers/component_test.js]: Should import this.

File: packages/ember-htmlbars/tests/helpers/component_test.js [hunk 7]
```diff
@@ -25,6 +27,49 @@ if (Ember.FEATURES.isEnabled('ember-htmlbars-component-helper')) {
...
+      set(view, "location", 'Loisaida');
+    });
+    equal(view.$().text(), 'yummy Loisaida arepas!', 'component was updated and re-rendered');
+
+    Ember.run(function() {
```
Comment (rwjblue) [inline packages/ember-htmlbars/tests/helpers/component_test.js]: Should import this.

**Comment by bagby** [comment]
@lukemelia it fixes that specific problem, but seems to have introduced another. Still investigating, I'll provide more info soon.

**Comment by krisselden** [comment]
we need a more general solution for update scheduling, the _level that is added to views was the start of aiding a priority queue for this, I think @mmun has some WIP continuing this effort.  I worried you guys are just going to be doing whack-a-mole going down this road.

**Comment by bagby** [comment]
Yes it feels a little whack-a-mole-ish, a very similar symptom is occurring a different place. It's like the pain moved from my elbow to my knee.

I have to put this down for now, if I get some time later I'll find a way of capturing it in a test.

**Comment by lukemelia** [comment]
@mmun @krisselden I'm not sure what to do here. Should we leave this PR open?

**Comment by mmun** [comment]
@lukemelia Yes. I apologize. I've been preoccupied with HTMLBars. I think that because it is behind a flag it is OK to merge if you are comfortable with it. We can review again before "go"ing it.

**Comment by rwjblue** [comment]
@mmun - The component helper is already 'go'ed. I don't disagree with your point (that we should merge), but we do need to circle around about this more before 1.11.0 final...

**Comment by lukemelia** [comment]
IMO, this is an edge case that would be good to fix once there is a clear path to do so, but personally, I would not let it block the release of 1.11.

**Comment by mmun** [comment]
Is there a regression test worth extracting out of this PR?

**Comment by rwjblue** [comment]
Great idea! This issue was closed with https://github.com/emberjs/ember.js/issues/11955 (a rebased version of this test). :dancers:

**Comment by mmun** [comment]
<3
